stations/views.py

```diff
@@ -165,3 +136,91 @@ def return_bike_to_station(self, request, *args, **kwargs):
             data=ReadBikeSerializer(bike).data,
             status=status.HTTP_201_CREATED,
         )
+
+
+class StationBlockedViewSet(
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    mixins.DestroyModelMixin,
+    GenericViewSet,
+):
+    queryset = Station.objects.filter(state=StationState.blocked)
+    serializer_class = StationSerializer
+
+    def handle_exception(self, exc):
+        if isinstance(exc, Http404):
+            return Response(
+                {"message": "Station not blocked."},
+                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+
+        return super().handle_exception(exc)
+
+    @restrict(UserRole.admin)
+    def create(self, request, *args, **kwargs):
+        """
+        Block a station.
+        Station id is provided in body.
+
+        Conditions:
+        - Station must exist
+        - Station must be currently working
+        """
+        station_id = request.data.get("id")
+        try:
+            station = Station.objects.get(id=station_id)
+        except Station.DoesNotExist:
+            return Response(
+                {"message": "Station not found."}, status=status.HTTP_404_NOT_FOUND
+            )
+        if station.state != StationState.working:
+            return Response(
+                {"message": "Station already blocked."},
+                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+
+        station.block()
+        self.cancel_all_reservations_at_station(station_id)
+        return Response(
+            {"id": str(station.id), "name": station.name},
+            status=status.HTTP_201_CREATED,
+        )
+
+    @restrict(UserRole.admin)
+    def destroy(self, request, *args, **kwargs):
+        """
+        Unblock a station.
+        Station id is provided in path.
+
+        Conditions:
+        - Station must exist
+        - Station must be currently blocked
+        """
+        try:
+            station = self.get_object()
+        except Station.DoesNotExist:
+            return Response(
+                {"message": "Station not found."}, status=status.HTTP_404_NOT_FOUND
+            )
+        if station.state != StationState.blocked:
+            return Response(
+                {"message": "Station not blocked."},
+                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+
+        station.unblock()
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    @restrict(UserRole.admin, UserRole.tech)
+    def list(self, request, *args, **kwargs):
+        return Response(
+            status=status.HTTP_200_OK,
+            data={"stations": StationSerializer(self.get_queryset(), many=True).data},
+        )
+
+    @staticmethod
+    def cancel_all_reservations_at_station(station_id):
+        bikes = Bike.objects.filter(station=station_id)
+        for bike in bikes:
+            if bike.status == BikeStatus.reserved:
+                bike.cancel_reservation()
```

Comment:
Mamy tutaj tzw. N+1 problem. To znaczy że robimy dodatkowe zapytanie per rower (np. 20 rowerów na staji = 20 zapytań zmainy statusu, 20 zapytań usunięcia rezerwacji).
Najlepiej byłoby przepisać całą tą metodę, zostaw proszę takie todo:
```
TODO(agorecki): optimize this to fix N+1 problem
```